odrive-motor/src/odriveCAN.py

```diff
@@ -0,0 +1,290 @@
+from typing import ClassVar, Mapping, Any, Dict, Optional, Tuple
+
+from typing_extensions import Self
+
+from viam.module.types import Reconfigurable
+from viam.proto.app.robot import ComponentConfig
+from viam.proto.common import ResourceName
+from viam.resource.base import ResourceBase
+from viam.resource.types import Model, ModelFamily
+
+from viam.components.motor import Motor
+from viam.logging import getLogger
+
+import odrive
+from odrive.enums import *
+from threading import Thread
+import asyncio
+import time
+import os
+from .utils import set_configs, find_baudrate, rsetattr, find_motor_configs
+
+import can
+import cantools
+import time
+
+db = cantools.database.load_file("odrive-cansimple.dbc")
+bus = can.Bus("can0", bustype="socketcan")
+
+LOGGER = getLogger(__name__)
+MINUTE_TO_SECOND = 60.0
+
+class OdriveCAN(Motor, Reconfigurable):
+    MODEL: ClassVar[Model] = Model(ModelFamily("viam-labs", "motor"), "odrive-canbus")
+    odrive_config_file: str
+    offset: float
+    baud_rate: str
+    odrv: Any
+    nodeID: int
+    torque_constant: float
+    current_soft_max: float
+
+    @classmethod
+    def new(cls, config: ComponentConfig, dependencies: Mapping[ResourceName, ResourceBase]) -> Self:
+        odriveCAN = cls(config.name)
+        odriveCAN.odrive_config_file = config.attributes.fields["odrive_config_file"].string_value
+        odriveCAN.nodeID = int(config.attributes.fields["canbus_node_id"].number_value)
+        odriveCAN.torque_constant = 1
+        odriveCAN.current_soft_max = 60
+        odriveCAN.offset = 0.0
+
+        try:
+            odriveCAN.odrv = odrive.find_any()
+            if odriveCAN.odrive_config_file != "":
+                set_configs(odriveCAN.odrv, odriveCAN.odrive_config_file)
+                rsetattr(odriveCAN.odrv, "axis0.config.can.node_id", odriveCAN.nodeID)
+                odriveCAN.torque_constant = find_motor_configs(odriveCAN.odrive_config_file, "torque_constant")
+                odriveCAN.current_soft_max = find_motor_configs(odriveCAN.odrive_config_file, "current_soft_max")
+        except TimeoutError:
+            LOGGER.warn("Could not set odrive configurations because no serial odrive connection was found.")
+
+        if config.attributes.fields["canbus_baud_rate"].string_value != "":
+            baud_rate = config.attributes.fields["canbus_baud_rate"].string_value
+            odriveCAN.baud_rate = baud_rate[0:len(baud_rate)-1] + "000"
+        elif odriveCAN.odrive_config_file != "":
+            baud_rate = find_baudrate(odriveCAN.odrive_config_file)
+            odriveCAN.baud_rate = str(baud_rate)
+        else:
+            odriveCAN.baud_rate = "250000"
+
+        os.system("sudo ip link set can0 down")
```

Comment:
this feels a bit wrong to be using sudo here. Would it be doable to use a package like pyroute2?  Otherwise would a nice error message pointing the user on how to setup can iface be bettre?

Reply:
looked into pyroute2 and was able to set can up and down, but I can't figure out how to set the bitrate. @npmenard if you have any suggestions on how to do that let me know, but if not I'll add instructions to the README about running these commands